[dynamo] use proxies to nn.Module in dynamo generated GraphModules

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120756

Fixes remaining refleaks found when debugging https://github.com/pytorch/pytorch/issues/119607, tests added in https://github.com/pytorch/pytorch/pull/120657.

Also fixes some tests that xfail: https://github.com/pytorch/pytorch/issues/120631 (not entirely sure why), but introduced tests now fail.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

williamwen42 committed:
Update

[ghstack-poisoned]

williamwen42 committed:
Update

[ghstack-poisoned]

williamwen42 committed:
Update

[ghstack-poisoned]

williamwen42 committed:
Update

[ghstack-poisoned]

williamwen42 committed:
Update

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] use proxies to nn.Module in dynamo generated GraphModules"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] use proxies to nn.Module in dynamo generated GraphModules"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] use proxies to nn.Module in dynamo generated GraphModules"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos commented on torch/_dynamo/utils.py:
```diff
@@ -2546,3 +2546,18 @@ class Invalid(dict):  # type: ignore[type-arg]
+
+# Super hacky - returns a proxy object for a nn.Module that references
+# to the original nn.Module, but is a distinct object.
+def nn_module_proxy(mod):
```
could you name this `wrapper` instead, due to the other uses of wrapper?

mlazos commented on torch/_dynamo/utils.py:
```diff
@@ -2546,3 +2546,18 @@ class Invalid(dict):  # type: ignore[type-arg]
+        return mod
+    try:
+        # best effort shallow copy
+        proxy = torch.nn.Module.__new__(torch.nn.Module)
```
Could you just use a WeakRef here instead? and do some routing to the underlying weakref

williamwen42 replied:
I attempted using weakrefs but discovered that a big difficulty in using weakrefs is that some downstream code (HOPs, export, backends) may require the actual module itself.

jansel commented on torch/_dynamo/utils.py:
```diff
@@ -2546,3 +2546,18 @@ class Invalid(dict):  # type: ignore[type-arg]
+        # best effort shallow copy
+        proxy = torch.nn.Module.__new__(torch.nn.Module)
+        proxy.__class__ = mod.__class__
+        proxy.__dict__ = mod.__dict__
```
Won't the parameters will still be held alive by this?

williamwen42 replied:
Parameters are lifted as inputs to GraphModules, so AFAIK a GraphModule should never have a Parameter as an attribute.

The point of this nn.Module wrapper/proxy is so that when the user deletes all their references to the original module, dynamo cache invalidation will work because the compiled GraphModule doesn't hold a reference to the original module.

williamwen42 committed:
Update on "[dynamo] use proxies to nn.Module in dynamo generated GraphModules"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
@jansel do you think it would be better if we explicitly write out which attributes of `nn.Module`s (e.g. params, modules, hooks, etc.) we want to copy over to the "proxy"/wrapper? I'm thinking it would be better for backend writers.

williamwen42 committed:
Update on "[dynamo] use proxies to nn.Module in dynamo generated GraphModules"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
@anijain2305 and I talked offline - we can keep the `__dict__` approach here since we have more control over the kinds of modules we're expecting from `output_graph.py` (i.e. inbuilt `torch.nn.Module`s, HOP-generated GraphModules)

williamwen42 committed:
Update on "[dynamo] use proxies to nn.Module in dynamo generated GraphModules"


Fixes remaining refleaks found when debugging https://github.com/pytorch/pytorch/issues/119607, tests added in https://github.com/pytorch/pytorch/pull/120657.

Also fixes https://github.com/pytorch/pytorch/issues/120631 (not entirely sure why)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang aakhundov

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge

yf225 commented on torch/_dynamo/utils.py:
```diff
@@ -2575,3 +2575,18 @@ class Invalid(dict):  # type: ignore[type-arg]
+    if isinstance(mod, torch.fx.GraphModule):
+        # Dynamo-generated GM's shouldn't contain user-created GM's
+        return mod
+    proxy = mod.__class__.__new__(mod.__class__)
```
hmm if module's `__new__` method expects additional arguments, maybe this won't work? e.g. per-parameter FSDP: https://github.com/pytorch/pytorch/blob/19d27a13ea052230d9fb565a5b82e683e28d1697/torch/distributed/_composable/fsdp/fully_shard.py#L142

williamwen42 replied:
What kind of Modules would these be? Do we expect the dynamo-generated GraphModule to contain them?

yf225 replied:
with inline_inbuilt_nn_modules, we won't hit this issue anymore which is nice haha
